### PR TITLE
MM-45899: Insights: least active channels

### PR DESCRIPTION
**Pull request by plant99**
#### Summary

This adds two new endpoints to get least active channels
`GET /api/v4/teams/:team_id/top/inactive_channels` and `GET /api/v4/users/me/top/inactive_channels`.

*Query Examples*:

Team insights
```
request
-------
curl -L -X GET 'http://localhost:8065/api/v4/teams/4f3rbqs5qibn5erzfazxgmssmh/top/inactive_channels?time_range=28_day&page=0&per_page=10' \ 
    -H 'X-Requested-With: XMLHttpRequest'\
    -H 'Authorization: Bearer gjaqjdbfhfre8cq9q4xq6ps5zy'

response
--------
{
  "has_next": true,
  "items": [
    {
      "id": "yftd1rq7spgtijok1zrziwk9co",
      "type": "O",
      "display_name": "Build failed",
      "name": "build-failed",
      "last_activity_at": 1658492886508,
      "participants": [
        "mb61ykxnxpg9zmkn3wjbqnafse",
        "z7pan8exabn3zr9shkaofh715w",
        "jqyz9tzkwprrx8zu46n74p7o3c"
      ]
    },
    {
      "id": "x7fdpjdhdififki47sjxqd86qa",
      "type": "O",
      "display_name": "Elastic Search unresponsive",
      "name": "elastic-search-unresponsive-c484",
      "last_activity_at": 1658485874884,
      "participants": [
        "mb61ykxnxpg9zmkn3wjbqnafse",
        "z7pan8exabn3zr9shkaofh715w"
      ]
    }
  ]
}
```

User insights
```
request
-------
curl -L -X GET 'http://localhost:8065/api/v4/teams/4f3rbqs5qibn5erzfazxgmssmh/top/inactive_channels?time_range=28_day&page=0&per_page=10' \
    -H 'X-Requested-With: XMLHttpRequest'\
    -H 'Authorization: Bearer gjaqjdbfhfre8cq9q4xq6ps5zy'

response
--------
{
  "has_next": false,
  "items": [
    {
      "id": "yftd1rq7spgtijok1zrziwk9co",
      "type": "O",
      "display_name": "Build failed",
      "name": "build-failed",
      "last_activity_at": 1658492886508,
      "participants": [
        "mb61ykxnxpg9zmkn3wjbqnafse",
        "z7pan8exabn3zr9shkaofh715w",
        "jqyz9tzkwprrx8zu46n74p7o3c"
      ]
    }]
}
```
#### Ticket Link

JIRA: https://mattermost.atlassian.net/browse/MM-45899
#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: api4/insights.go [hunk 1]
```diff
@@ -340,6 +344,106 @@ func getTopThreadsForUserSince(c *Context, w http.ResponseWriter, r *http.Reques
...
+		return
+	}
+
+	js, jsonErr := json.Marshal(topChannels)
+	if jsonErr != nil {
+		c.Err = model.NewAppError("getTopInactiveChannelsForTeamSince", "api.marshal_error", nil, jsonErr.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write(js)
```
Comment (agnivade) [inline api4/insights.go:391]: Hey Shivasish, apologies for the drive-by comment. Just one small thing - usually we follow the convention:

```go
if err := json.NewEncoder(w).Encode(topChannels); err != nil {
  c.Logger.Warn("Error while writing top channels", mlog.Err(err))
}
```

This does the marshalling and writing the response in a single line, and also captures the error from writing the bytes to the network.

Comment (plant99) [reply api4/insights.go:391]: Thanks for the suggestion, this seems a lot better. I'll apply this change to wherever we're doing this separately.

Comment (plant99) [reply api4/insights.go]: Fixed, thanks! Also added a ticket to fix it in other endpoints, https://mattermost.atlassian.net/browse/MM-46331

**Comment by BenCookie95** [comment]
@plant99 For the participant array I only need there IDs so an array of ID strings is fine

**Review by mkraft — COMMENTED** [review]
Could you please include examples of the queries in the description? Each new endpoint makes two queries so I guess that's 4 query examples.

File: store/sqlstore/channel_store.go [hunk 2]
```diff
@@ -4319,6 +4319,198 @@ func (s SqlChannelStore) GetTopChannelsForUserSince(userID string, teamID string
...
+			LastActivityAt
+		FROM
+			((SELECT
+				Posts.ChannelId AS ID,
+				'O' AS Type,
+				PublicChannels.DisplayName AS DisplayName,
+				PublicChannels.Name AS Name,
+				PublicChannels.TeamId AS TeamID,
+				count(Posts.Id) AS MessageCount,
+				PublicChannels.DeleteAt AS DeleteAt,
```
[the remark this answers is not in the document]
Comment (plant99) [reply store/sqlstore/channel_store.go:4348]: Thanks for pointing this out, I think we should add `Channels.DeleteAt` to the WHERE clause in both queries.

[the remark this answers is not in the document]
Comment (plant99) [reply store/sqlstore/channel_store.go:4348]: Please correct me if we decided to include archived channels in top inactive, I think that'd be very less beneficial.

File: store/sqlstore/channel_store.go [hunk 3]
```diff
@@ -4319,6 +4319,232 @@ func (s SqlChannelStore) GetTopChannelsForUserSince(userID string, teamID string
...
+	postgresPropQuery := `AND (Posts.Props ->> 'from_bot' IS NULL OR Posts.Props ->> 'from_bot' = 'false') AND (Posts.Props ->> 'from_webhook' IS NULL OR Posts.Props ->> 'from_webhook' = 'false')`
+	mySqlPropsQuery := `AND (JSON_EXTRACT(Posts.Props, '$.from_bot') IS NULL OR JSON_EXTRACT(Posts.Props, '$.from_bot') = 'false') AND (JSON_EXTRACT(Posts.Props, '$.from_webhook') IS NULL OR JSON_EXTRACT(Posts.Props, '$.from_webhook') = 'false')`
+
+	query := `
+		SELECT
+			ID,
+			Type,
+			DisplayName,
+			Name,
+			TeamID,
```
[the remark this answers is not in the document]
Comment (plant99) [reply store/sqlstore/channel_store.go:4337]: @BenCookie95 do you remember using TeamID in `TopChannels`?

File: store/sqlstore/channel_store.go [hunk 4]
```diff
@@ -4319,6 +4319,232 @@ func (s SqlChannelStore) GetTopChannelsForUserSince(userID string, teamID string
...
+				Posts.DeleteAt = 0
+				AND Posts.CreateAt > ?
+				AND Posts.Type = ''`
+	args = []any{since}
+
+	if s.DriverName() == model.DatabaseDriverMysql {
+		query += mySqlPropsQuery
+	} else if s.DriverName() == model.DatabaseDriverPostgres {
+		query += postgresPropQuery
+	}
```
[the remark this answers is not in the document]
Comment (plant99) [reply store/sqlstore/channel_store.go:4363]: I forgot to add this, thanks for pointing it out. Will do!

File: store/sqlstore/channel_store.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenCookie95) [reply store/sqlstore/channel_store.go:4337]: We don't need TeamID on the frontend so feel free to remove it

**Comment by plant99** [comment]
> @plant99 For the participant array I only need there IDs so an array of ID strings is fine

Also @BenCookie95 , please note I've fixed this.

File: store/sqlstore/channel_store.go [hunk 5]
```diff
@@ -4319,6 +4319,199 @@ func (s SqlChannelStore) GetTopChannelsForUserSince(userID string, teamID string
...
+			GROUP BY
+				Posts.ChannelId,
+				Channels.Type,
+				Channels.DisplayName,
+				Channels.Name)) AS A
+		ORDER BY
+			MessageCount ASC,
+			Name ASC
+		LIMIT ?
+		OFFSET ?`
```
Comment (mkraft) [inline store/sqlstore/channel_store.go:4394]: Can we combine all three `args = append` variable assignments and the string concatenations of `query`?

**Review by mkraft — CHANGES_REQUESTED** [review]

File: store/sqlstore/channel_store.go (hunk 5 above)
Comment (plant99) [reply store/sqlstore/channel_store.go:4388]: Thanks Martin, fixed it.

Had kept it for better readability. But I see that 7 arguments doesn't make a lot of difference.

**Comment by plant99** [comment]
@mkraft updated description with sample queries. But they'd be two, right?

**Review by mkraft — APPROVED** [review]

**Review by BenCookie95 — APPROVED** [review]
LGTM!

**Review by furqanmlk — APPROVED** [review]
